File: 0747-min-cost-climbing-stairs/0747-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        
        @cache
        def dp(i: int) -> int:
            if i >= len(cost):
                return 0
            
            return min(dp(i + 1), dp(i + 2)) + cost[i]
        
        return min(dp(0), dp(1))

File: 0747-min-cost-climbing-stairs/0747-min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();

        vector<int> dp(n, -1);

        function<int(int)> find_cost_from_stair = [&] (int i) -> int {
            if (i >= n) return 0;
            if (dp[i] != -1) return dp[i];
            
            dp[i] = cost[i] + min(find_cost_from_stair(i + 1), find_cost_from_stair(i + 2));
            return dp[i];
        };

        return min(find_cost_from_stair(0), find_cost_from_stair(1));


    }
};